controllers/defaults/placements.go

```diff
@@ -45,17 +45,71 @@ var (
 			Tolerations: []corev1.Toleration{
 				getOcsToleration(),
 			},
+			// The first TSC is a hard constraint which restricts the movement of OSDs between failure domain
+			// The second TSC is a soft constraint which restricts the movement of OSDs between hosts
+			// The topology key in the first TSC is set to empty string, it should be updated to the failure domain key
 			TopologySpreadConstraints: []corev1.TopologySpreadConstraint{
-				getTopologySpreadConstraintsSpec(1, []string{osdLabelSelector}),
+				{
+					MaxSkew:           1,
+					TopologyKey:       "",
```

Comment:
What if `flexibleScaling: true`? Seems like in that case we should change the host-based TSC to required and not have a second TSC.

Reply:
Hm, right. We have basically 3 types of failure domains & we need to think of proper spec for each time.
As far as I have understood I am summarizing the requirement below, do correct me if I am wrong anywhere.

```
If the failure domain is zone
We need soft constraints on hostname key, We don't want to put hard constraint on failure domain level.

For rack failure domain
We need hard constraint on rack level, with soft constraint on hostname level. 

For hostname failure domain(flexible scaling)
We need hard constraint on hostname level
```

Reply:
> If the failure domain is zone
> We need soft constraints on hostname key, We don't want to put hard constraint on failure domain level.

Why wouldn't we put a hard constraint on the zone level? I thought we always needed a hard constraint for the top level failure domain.

Reply:
That was my expectation too, but if you see here in the old code(which was written around 4 years back), Adding hard constraint on zone level is specifically avoided which I am not sure why

https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/cephcluster.go#L815

Reply:
I found this little discussion where you raised the same question https://github.com/red-hat-storage/ocs-operator/pull/1043#discussion_r573147451

Reply:
My opinion would be keeping a consistent behaviour in this way
1. Hard constraint on the failure domain level
2. soft constraint on the hostname level

Reply:
Wow nice old thread! :) Ok sounds good to keep it as previously with the soft constraint on zones as described there.